### PR TITLE
8341755: Optimize argNames in InnerClassLambdaMetafactory

### PR DESCRIPTION
**Pull request by wenshao**
A simple optimization that eliminates the allocation of the MethodTypeDescImpl object when parameterCount is equal to 0 and eliminates the allocation of argNames when parameterCount is equal to 1

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341755](https://bugs.openjdk.org/browse/JDK-8341755): Optimize argNames in InnerClassLambdaMetafactory (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) Review applies to [191bba08](https://git.openjdk.org/jdk/pull/21399/files/191bba086a54ac88d8ffa6b3c0ce5b2ad018cfaa)

### Contributors
 * Chen Liang `<liach@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21399/head:pull/21399` \
`$ git checkout pull/21399`

Update a local copy of the PR: \
`$ git checkout pull/21399` \
`$ git pull https://git.openjdk.org/jdk.git pull/21399/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21399`

View PR using the GUI difftool: \
`$ git pr show -t 21399`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21399.diff">https://git.openjdk.org/jdk/pull/21399.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21399#issuecomment-2400181184)

### CODE REVIEW

**Comment by liach** [comment]
`argNames[i]` is always `"arg$" + (i + 1)`. Maybe we can change all `argNames[x]` usage to a method like:
```java
public static String argName(int i) {
    if (i < 8)
        return ARG_NAME_CACHE[i];
    return "arg$" + (i + 1);
}
```

**Comment by liach** [comment]
Can we change the name to "Optimize argNames in InnerClassLambdaMetafactory"?

**Review by liach — APPROVED** [review]
Looks good. Since I have contributed I think we need another review, such as @cl4es.

/reviewers 2 reviewer

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java [hunk 1]
```diff
@@ -174,18 +171,31 @@ public InnerClassLambdaMetafactory(MethodHandles.Lookup caller,
...
+            constructorTypeDesc = MTD_void;
+        }
+        this.argDescs = argDescs;
+        this.constructorTypeDesc = constructorTypeDesc;
+    }
+
+    private static String argName(int i) {
+        String argName = i < ARG_NAME_CACHE_SIZE ? ARG_NAME_CACHE[i] : null;
+        if (argName == null) {
+            argName = "arg$".concat(Integer.toString(i + 1));
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:193]: With `Integer.toString` in the mix perhaps just as well keep this as a standard concat?
```suggestion
            argName = "arg$" + (i + 1);
```

**Review by cl4es — APPROVED** [review]
Looks reasonable

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:193]: New bytecode in last commit's `computeArgName`:
```
         0: ldc           #186                // String arg$
         2: iload_0
         3: iconst_1
         4: iadd
         5: invokestatic  #188                // Method java/lang/Integer.toString:(I)Ljava/lang/String;
         8: invokevirtual #193                // Method java/lang/String.concat:(Ljava/lang/String;)Ljava/lang/String;
        11: areturn

```
Old bytecode in constructor:
```
       222: iload         11
       224: new           #150                // class java/lang/StringBuilder
       227: dup
       228: invokespecial #152                // Method java/lang/StringBuilder."<init>":()V
       231: ldc           #155                // String arg$
       233: invokevirtual #157                // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
       236: iload         11
       238: iconst_1
       239: iadd
       240: invokevirtual #161                // Method java/lang/StringBuilder.append:(I)Ljava/lang/StringBuilder;
       243: invokevirtual #164                // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
       246: aastore
       247: aload_0
```
I believe using concat is cleaner.

Comment (cl4es) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:193]: Irrelevant; idiomatic concatenation is more concise in source and easier to read, comprehend and check, so should generally be preferred. That the legacy javac concat translatiom is lousy for constant + primitive is a different matter.

**Review by liach — APPROVED** [review]

**Review by cl4es — APPROVED** [review]

**Review by liach — APPROVED** [review]
Trivial merge.

/reviewers 1 reviewer

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java [hunk 2]
```diff
@@ -174,18 +171,31 @@ public InnerClassLambdaMetafactory(MethodHandles.Lookup caller,
...
-            argNames = EMPTY_STRING_ARRAY;
             argDescs = EMPTY_CLASSDESC_ARRAY;
+            constructorTypeDesc = MTD_void;
+        }
+        this.argDescs = argDescs;
+        this.constructorTypeDesc = constructorTypeDesc;
+    }
+
+    private static String argName(int i) {
+        String argName = i < ARG_NAME_CACHE_SIZE ? ARG_NAME_CACHE[i] : null;
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:191]: Just an idea, but perhaps could constant init `ARG_NAME_CACHE = { "arg$0", "arg$1", .. , "arg$7" };` and simplify this?

**Review by liach — APPROVED** [review]
Back to precomputed constants :)

**Comment by wenshao** [comment]
> Back to precomputed constants :)

Manual shift computation! The code is also simplified based on precomputed constants.

**Pull request closed without merging**